File: csharp/TransactionAnalyzer.Lib/Transaction.cs
#nullable enable
using System;

namespace TransactionAnalyzer.Lib
{
    public record Transaction
    {
        public string Id { get; init; }
        public DateTime TransactedAt { get; init; }
        public decimal Amount { get; init; }
        public string MerchantName { get; init; }
        public TransactionType Type { get; init; }
        public string? RelatedTransactionId { get; init; }

        // public Transaction(
        //     string id,
        //     DateTime transactedAt,
        //     decimal amount,
        //     string merchantName,
        //     TransactionType type,
        //     string? relatedTransactionId=""
        // ) => (Id,
        //         TransactedAt,
        //         Amount,
        //         MerchantName,
        //         Type,
        //         RelatedTransactionId)
        //     = (id,
        //         transactedAt,
        //         amount,
        //         merchantName,
        //         type,
        //         relatedTransactionId);


    }
}

File: transaction-analyzer/php/src/Transaction.php
<?php

namespace TransactionAnalyser;

use Brick\DateTime\LocalDateTime;
use Brick\Math\BigDecimal;

class Transaction
{

    function __construct(
        private string $id,
        private LocalDateTime $transactedAt,
        private BigDecimal $amount,
        private string $merchantName,
        private TransactionType $type,
        private string|null $relatedTransactionId = null)
    {
    }

    function getId(): string
    {
        return $this->id;
    }

    /**
     * @return LocalDateTime
     */
    public function getTransactedAt(): LocalDateTime
    {
        return $this->transactedAt;
    }

    /**
     * @return BigDecimal
     */
    public function getAmount(): BigDecimal
    {
        return $this->amount;
    }

    public function getMerchantName(): string
    {
        return $this->merchantName;
    }

    public function getType(): TransactionType
    {
        return $this->type;
    }

    public function getRelatedTransactionId(): string|null
    {
        return $this->relatedTransactionId;
    }

}
